src/hash_to_field.cairo

```diff
@@ -0,0 +1,311 @@
+        assert [msg_hash_train + 33] = 0x57555F52;
+        assert [msg_hash_train + 34] = 0x4F5F4E55;
+        assert [msg_hash_train + 35] = 0x4C5F2B;  // DST + DST.len
+
+        // Compute the initial hash (b_0)
+        let (msg_hash) = SHA256.hash_bytes(msg_hash_train, 111 + msg_bytes_len);  // 64b z_pad + msg_bytes_len + 2b block_size, 0x0 ++ 43b dst + 1b dst_len
+
+        // Prepare input for the first block hash (b_1)
+        let (hash_args: felt*) = alloc();
+        memcpy(dst=hash_args, src=msg_hash, len=B_IN_FELTS);
+        let one_dst_prime = ONE_DST_PRIME();
+        memcpy(dst=hash_args + 8, src=one_dst_prime, len=12);
+
+        // Compute the first block hash (b_1)
+        let (hash_1) = SHA256.hash_bytes(hash_args, 77);  // 32b msg + 1b 0x1 + 43b dst + 1b dst_len
+
+        // Create hash_train and copy first hash. The hash_train contains all
+        let (hash_train: felt*) = alloc();
+        memcpy(dst=hash_train, src=hash_1, len=B_IN_FELTS);
+
+        with one_dst_prime {
+            expand_msg_xmd_inner(msg_hash=msg_hash, hash_train=hash_train, ell=ell, index=0);
+        }
+
+        // Copy the result. Potentially remove this
+        let (result: felt*) = alloc();
+        memcpy(dst=result, src=hash_train, len=n_bytes / 4);
+
+        return (result=result);
+    }
+
+    // Inner recursive function for expand_msg_xmd
+    func expand_msg_xmd_inner{
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+        sha256_ptr: felt*,
+        pow2_array: felt*,
+        one_dst_prime: felt*,
+    }(msg_hash: felt*, hash_train: felt*, ell: felt, index: felt) {
+        alloc_locals;
+
+        if (index == ell) {
+            return ();
+        }
+
+        // XOR the initial hash (b_0) with the previous block hash
+        let xored = _xor_hash_segments(msg_hash, hash_train + index * B_IN_FELTS);
+
+        // Prepare the input for the next block hash:
+        // (b_0 XOR b_i) || i+1 || DST || len(DST)
+        assert [xored + 8] = [one_dst_prime] + 0x01000000 * (index + 1);
+        memcpy(dst=xored + 9, src=one_dst_prime + 1, len=11);
+
+        // Compute the next block hash
+        let (hash) = SHA256.hash_bytes(xored, 77);
+        // Store the new block hash in the output array
+        memcpy(dst=hash_train + (index + 1) * B_IN_FELTS, src=hash, len=B_IN_FELTS);
+
+        return expand_msg_xmd_inner(
+            msg_hash=msg_hash, hash_train=hash_train, ell=ell, index=index + 1
+        );
+    }
+
+    // Hashes a message to a field element
+    // Implementation of algorithm from:
+    // https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-11#section-5.3
+    //
+    // Parameters:
+    // - msg: a byte string containing the message to hash, chunked in big-endian 4-bytes
+    // - msg_bytes_len: length of the message in bytes
+    // - count: the number of field elements to output
+    //
+    // Returns:
+    // - fields: an array of field elements [u_0, ..., u_(count - 1)]
```

Comment:
Specify the output is an array of pointers of length curve_M and make curve_M a parameter to this function (name ext_degree)

So that we can re-use it for hash_to_g1 as well.